packages/tasit-sdk/src/Decentraland.test.js

```diff
@@ -249,393 +314,657 @@ describe("Decentraland", () => {
 
-          const allowance = await mana.allowance(
-            ephemeralAddress,
-            MARKETPLACE_ADDRESS
-          );
+            const confirmationListener = async () => {
+              await expectExactEtherBalances(
+                provider,
+                [toAddress],
+                [SMALL_AMOUNT]
+              );
 
-          expect(`${allowance}`).to.equal(`${manaAmountForShopping}`);
-        });
+              done();
+            };
+
+            const errorListener = error => {
+              action.off("error");
+              done(error);
+            };
+
+            action.once("confirmation", confirmationListener);
+            action.on("error", errorListener);
 
-        it("should buy an estate", async () => {
-          const {
-            assetId,
-            nftAddress,
-            seller,
-            priceInWei,
-            expiresAt,
-          } = estateForSale;
+            action.send();
+          })();
+        });
 
-          await checkAsset(estate, mana, estateForSale, ephemeralAddress);
+        beforeEach("onboarding - funding ephemeral wallet with MANA", done => {
+          (async () => {
+            // Global hooks don't run between same level hooks
+            await mineBlocks(provider, 1);
+
+            const toAddress = ephemeralAddress;
+
+            const action = gnosisSafe.transferERC20(
+              MANA_ADDRESS,
+              toAddress,
+              manaAmountForShopping
+            );
+
+            const confirmationListener = async message => {
+              await expectExactTokenBalances(
+                mana,
+                [toAddress],
+                [manaAmountForShopping]
+              );
+              done();
+            };
+
+            const errorListener = error => {
+              action.off("error");
+              done(error);
+            };
+
+            action.once("confirmation", confirmationListener);
+            action.on("error", errorListener);
+
+            action.send();
+          })();
+        });
 
-          await expectExactTokenBalances(estate, [ephemeralAddress], [0]);
+        beforeEach("onboarding - approving Marketplace to spend MANA", done => {
+          (async () => {
+            // Global hooks don't run between same level hooks
+            await mineBlocks(provider, 1);
+
+            mana.setWallet(ephemeralWallet);
+            const action = mana.approve(
+              MARKETPLACE_ADDRESS,
+              manaAmountForShopping
+            );
+
+            const confirmationListener = async message => {
+              const allowance = await mana.allowance(
+                ephemeralAddress,
+                MARKETPLACE_ADDRESS
+              );
+
+              expect(`${allowance}`).to.equal(`${manaAmountForShopping}`);
+              done();
+            };
+
+            const errorListener = error => {
+              action.off("error");
```

Comment:
Any reason this is needed? Maybe as part of test cleanup all listeners should be removed?

Reply:
Yes, we can move to share `action` var between all test cases and having a cleanup (same as Contract.test.js), but I like the way that it is because reflects better how it's should be used on a real case. Thoughts?

Reply:
Hmm, so you're saying the user in reality should remove an error listener the moment they get an error? Why? What if there were some second relevant error?

Reply:
Good point. This is the way that is working now. So, maybe we what that the error listener works always on the "`once` mode"?

Reply:
Hmm wait, I'm still confused - I think we may disagree?

I think users will usually want to listen to errors forever, with some exceptions.

Are you saying that users will usually only want to listen to the first error?

Reply:
I think we maybe will have both situations. Examples:
- Keep `error` listener running and turn it off when the action was successful;
- Turn the `error` listener off if an error occurs (for some reason).
Yes, not make sense unsubscribe after an error emission.


So, answering your question again:

> Any reason this is needed? Maybe as part of test cleanup all listeners should be removed?

You call, I don't have a strong opinion but I like to not have to use global var and having a complete isolated test case.

Reply:
I had a little confusion stemming from picturing how users would use this feature vs. how we want to architect the test suite.

I agree that not having to use a global var in the test suite is nice. So we don't need to create any test cleanup setup that handles cleanup for listeners from all tests.

And I agree that users or our test suite might want to do like you described in both situations above.

What's the problem that we would run into if we kept the error listener subscribed unnecessarily here? Just memory leak at scale? Or intermittent test failures?

Reply:
Long story short - we can keep this line of code as-is, but I'm still curious about the last question for future reference.

Reply:
> What's the problem that we would run into if we kept the error listener subscribed unnecessarily here? Just memory leak at scale? Or intermittent test failures?

Mocha can be stuck (failing test suite by timeout), not sure but non-determinism can occur since we are using the same provider for all tests on the same test suite.

Reply:
That makes sense. Thanks. Resolving this.